lib/plugins/digging.js

```diff
@@ -13,22 +15,104 @@ function inject (bot) {
     if (block === null || block === undefined) {
       throw new Error('dig was called with an undefined or null block')
     }
+    if (!digFace || typeof digFace === 'function') {
+      digFace = 'auto'
+    }
 
     if (bot.targetDigBlock) bot.stopDigging()
 
+    let diggingFace = 1 // Default (top)
+
     if (forceLook !== 'ignore') {
-      await bot.lookAt(block.position.offset(0.5, 0.5, 0.5), forceLook)
+      if (digFace?.x || digFace?.y || digFace?.z) {
+        // Determine the block face the bot should mine
+        if (digFace.x) {
+          diggingFace = digFace.x > 0 ? BlockFace.EAST : BlockFace.WEST
+        } else if (digFace.y) {
+          diggingFace = digFace.y > 0 ? BlockFace.TOP : BlockFace.BOTTOM
+        } else if (digFace.z) {
+          diggingFace = digFace.z > 0 ? BlockFace.SOUTH : BlockFace.NORTH
+        }
+        await bot.lookAt(block.position.offset(0.5, 0.5, 0.5).offset(digFace.x * 0.5, digFace.y * 0.5, digFace.z * 0.5), forceLook)
+      } else if (digFace === 'raycast') {
+        // Check faces that could be seen from the current position. If the delta is smaller then 0.5 that means the bot cam most likely not see the face as the block is 1 block thick
+        // this could be false for blocks that have a smaller bounding box then 1x1x1
+        const dx = bot.entity.position.x - block.position.x + 0.5
+        const dy = bot.entity.position.y - block.position.y - 0.5 + bot.entity.height // -0.5 because the bot position is calculated from the block position that is inside its feet so 0.5 - 1 = -0.5
+        const dz = bot.entity.position.z - block.position.z + 0.5
+        // Check y first then x and z
+        const visibleFaces = {
+          y: Math.sign(Math.abs(dy) > 0.5 ? dy : 0),
+          x: Math.sign(Math.abs(dx) > 0.5 ? dx : 0),
+          z: Math.sign(Math.abs(dz) > 0.5 ? dz : 0)
+        }
+        const validFaces = []
+        outerLoop:
+        for (const i in visibleFaces) {
+          if (!visibleFaces[i]) {
+            // skip as this face is not visible
+            continue
+          }
+          const targetPos = block.position.offset(0.5 + (i === 'x' ? visibleFaces[i] * 0.5 : 0), 0.5 + (i === 'y' ? visibleFaces[i] * 0.5 : 0), 0.5 + (i === 'z' ? visibleFaces[i] * 0.5 : 0))
+          const startPos = bot.entity.position.offset(0, bot.entity.height, 0)
+          const rayIterator = new RaycastIterator(startPos, targetPos.clone().subtract(startPos).normalize(), 5) // overshoot a bit for more reliable raycasting
+          let rayBlock = rayIterator.next()
+          let last = rayBlock
+          while (rayBlock) {
+            if (rayBlock.x === block.position.x && rayBlock.y === block.position.y && rayBlock.z === block.position.z) {
+              if (last) {
+                console.info('Valid face', targetPos, last.face)
+                // is the face value switched for rayIterator.next() ? Should be the face it entered not exited ?
```

Comment:
To get the right face of the block I want to hit I had to use the last block I entered with the raycast. Im not sure if this is a bug in the Raycasting iterator or if my code is wrong in some way?